### PR TITLE
Round-End Statistics Expansion

### PR DESCRIPTION
**Pull request by ghost**
- Statistics displayed at round end now use a new system
- Text formatting has been changed for displayed statistics
- New stats are now displayed at round end, such as total player deaths and how many times the clown honked their horn

Post-Undertale fervor is a hell of a drug.

### CODE REVIEW

**Comment by Incoming5643** [comment]
I'm all for the occasional semi hidden reference, but one that appears at the end of every round about a game that didn't even win a "the game awards" award seems more than a bit much.

File: code/controllers/subsystem/ticker.dm [hunk 1]
```diff
@@ -489,3 +473,31 @@ var/datum/subsystem/ticker/ticker
...
+	if(station_evacuated)
+		world << "<br><b>Evacuation Rate:</b> [num_escapees] ([round((num_escapees/joined_player_list.len)*100, 0.1)]%"
+
+	//Announce who killed the most people
+	var/most_kills = -1
+	for(var/mob/M in player_list)
+		if(M.client && M.client.murders > most_kills)
+			chara = M.client
+			most_kills = M.client.murders
+	var/list/fluff_messages = list("<font color='red'>* In my way.</span>", "Had the highest LOVE", "Had the most EXP")
```
Comment (Iamgoofball) [inline code/controllers/subsystem/ticker.dm]: can we not

**Comment by Incoming5643** [comment]
Also yeah glorifying killing people can lead to terrible players.

Hell technically a nonantag can win it.

**Comment by Zilenan91** [comment]
Yeah no displaying the highest kills by a single person would do nothing but harm the server. Please remove that.

**Comment by francinum** [comment]
Showing the player with the most kills is a terrible idea.

**Comment by Xhuis** [comment]
I'll replace the Most Kills stat with the Fewest Kills stat.
As for the references, they're staying unless a maintainer wants 'em out.

**Comment by Incoming5643** [comment]
I don't exactly see the point of fewest kills either, it's just gonna be a tie for 0

**Comment by Zilenan91** [comment]
The references are probably fine, but how will you track fewest kills without the amount of people stated being massive? In prime time we got like 80-90 people, so, being generous, what if 40 or so people don't get a single kill?

**Comment by Incoming5643** [comment]
Alright let me be less nice.

Fuck this flavor of the month nonsense.

**Comment by Xhuis** [comment]
The way I figured, violence is barely avoidable in this game, but your points are convincing. I'll change them in the morning

**Comment by Incoming5643** [comment]
statistically for no one in the game to have zero kills to their name requires every player to have died by someone elses hand.

**Comment by newfren** [comment]
Incoming you're forgetting about cloning, it introduces more possible kills to the round without increasing the amount of players.

It'll still be 20 people tying for 0 though.

File: code/controllers/subsystem/ticker.dm (hunk 1 above)
Comment (Iamgoofball) [inline code/controllers/subsystem/ticker.dm]: oh to clarify i'm shitting on the fluff messages
i'm all for a kill counter

**Comment by Iamgoofball** [comment]
hi I'm not talking about the kill stat

just the fluff messages

**Comment by Xhuis** [comment]
I'll probably make killcount statistics a config option when I figure out how, so I'm leaving the code in for now. In the meantime, I added three more statistics:
* Total Player Deaths - the amount of times a player-controlled mob has died
* How Many Times The Clown Honked - the amount of times a legitimate clown has honked a bike horn
* Total Slips - the amount of times a mob has slipped

**Comment by Jacquerel** [comment]
Can we have a "player who died and was cloned the most times" statistic?

**Comment by bgobandit** [comment]
can you port the thing from Facepunch where this shows up in a separate window, or even better, make that a prefs option

File: code/modules/mob/living/death.dm [hunk 2]
```diff
@@ -35,6 +35,12 @@
 	flick(flick_name, animate)
 
 /mob/living/death(gibbed)
+	if(client && ismob(lastattacker)) //If they were murdered, increase the murderer's client kill count! We CoD now.
+		var/mob/M = lastattacker
+		if(M.client)
+			M.client.murders++ //Your LOVE increased.
```
Comment (RemieRichards) [inline code/modules/mob/living/death.dm]: this is false on so many levels.
Hit someone once and then have them killed by the environment = YOU MURDERED THEM!

Even if their death happened 5 hours later while you were logged off.

File: code/controllers/subsystem/statistics.dm [hunk 3]
```diff
@@ -0,0 +1,7 @@
+//This isn't actually a subsystem. It exists solely to hold a game's round statistics, such as deaths.
```
Comment (AnturK) [inline code/controllers/subsystem/statistics.dm]: Then why is it in here

**Comment by AnturK** [comment]
Anyway, i like statistics so this is nice.

**Comment by Shadowlight213** [comment]
:-1: on kill statistics. Silly stuff or stuff like cargo points is much better.

File: code/controllers/subsystem/statistics.dm (hunk 3 above)
Comment (Xhuis) [inline code/controllers/subsystem/statistics.dm]: Didn't know where else to put it :(

File: code/modules/mob/living/death.dm (hunk 2 above)
Comment (Xhuis) [inline code/modules/mob/living/death.dm]: Do you know of a better way? I would appreciate it

Comment (RemieRichards) [inline code/modules/mob/living/death.dm]: There really isn't one.

**Comment by PKPenguin321** [comment]
players with the most deaths would be funny
ive died like 7 times in one round before

File: code/controllers/subsystem/statistics.dm (hunk 3 above)
Comment (optimumtact) [inline code/controllers/subsystem/statistics.dm]: just put it as a global var datum initialised at world start or attach it to world var

File: code/controllers/subsystem/ticker.dm [hunk 4]
```diff
@@ -489,3 +473,45 @@ var/datum/subsystem/ticker/ticker
...
+
+	//Generic stats: time, integrity, population...
+	world << "<br><b>Shift Duration:</b> [round(world.time / 36000)]:[add_zero("[world.time / 600 % 60]", 2)]:[world.time / 100 % 6][world.time / 100 % 10]"
+	world << "<br><b>Station Integrity:</b> [mode.station_was_nuked ? "<font color='red'>Destroyed</font>" : "[station_integrity]%"]"
+	world << "<br><b>Total Population:</b> [joined_player_list.len]"
+	world << "<br><b>Survival Rate:</b> [num_survivors] ([round((num_survivors/joined_player_list.len)*100, 0.1)]%)"
+	if(station_evacuated)
+		world << "<br><b>Evacuation Rate:</b> [num_escapees] ([round((num_escapees/joined_player_list.len)*100, 0.1)]%"
+
+	/* Uncomment this block to enable announcement of the people with the highest and lowest kill count
```
Comment (optimumtact) [inline code/controllers/subsystem/ticker.dm]: make this a a config setting, don't leave it commented out

downstreams may want it.

Comment (Incoming5643) [inline code/controllers/subsystem/ticker.dm]: We've already proven that this will return a huge tie in nearly every case

**Comment by Dorsisdwarf** [comment]
+1 most deaths per round

**Comment by GunHog** [comment]
Death count and most deaths are both easier to track and amusing without the controversy of appearing to promote murderbone. (Although, in my opinion, murderbone is good, and adding nerfs to prevent muderbone is bad)

**Comment by Robustin** [comment]
Displaying a kill count is hilarious and I support it. Gives something for people to talk about, helps them remember a particular occasion, etc. Just only put it in modes where it makes sense to glorify bodycounts (Gang War, Traitor, Ling, Nukeops maybe, Malf, Bleb, Wizard).

**Comment by Incoming5643** [comment]
you don't really need kill counts in single antagonist modes though, since one can reasonably gleam that most of the people missing  from the survival stat died to the primary antag.

File: code/modules/mob/living/death.dm (hunk 2 above)
Comment (Iamgoofball) [inline code/modules/mob/living/death.dm]: with the probably quite literally 1500 ways to kill people in SS13 there is no way we can track who killed who without manual log searching.

**Comment by Zilenan91** [comment]
The last thing I want to see is people sucking murderboner dick.

I mean, it's cool conceptually, but for shit like traitor/ling/gang/rev or whatever it'd just be awful because those modes would no longer be about how much fun everyone had, but the body count that piled up.

It'd be cool in wizard though.

**Comment by Dorsisdwarf** [comment]
Killcount was removed and commented out, people. Its probbably not coming back.

File: code/controllers/subsystem/statistics.dm [hunk 5]
```diff
@@ -0,0 +1,7 @@
...
+
+var/client/chara = null //The client who killed the most people
+var/client/frisk = null //The client who killed the least people
```
Comment (FlavoredCactus) [inline code/controllers/subsystem/statistics.dm]: Undertale memes aaaaa why?

File: code/controllers/subsystem/ticker.dm [hunk 6]
```diff
@@ -489,3 +473,45 @@ var/datum/subsystem/ticker/ticker
...
+	var/list/fluff_messages = list("<font color='red'>* In my way.</span>", "Had the highest LOVE", "Had the most EXP")
+	world << "<br><b>[pick(fluff_messages)]:</b> [chara] <i>([chara.murders] kills)</i>"
+
+	//Announce who killed the least people:
+	var/least_kills = chara.murders
+	for(var/mob/M in player_list)
+		if(M.client && M.client.murders < least_kills)
+			frisk = M.client
+			least_kills = frisk.murders
+	fluff_messages = list("Has a stupid saccharine schtick", "Had the lowest LOVE", "Had the least EXP")
```
Comment (FlavoredCactus) [inline code/controllers/subsystem/ticker.dm]: Oh the memes extend beyond just var names and instead actual messages. these should be more general rather then references to undertale.

File: code/controllers/subsystem/statistics.dm (hunk 5 above)
Comment (Giacom) [inline code/controllers/subsystem/statistics.dm]: I would not recommend putting the references in the var names themselves, as not everyone has played Undertale and will understand their purpose without having to look at the comment near the definition. It just makes it unreadable for some.

File: code/controllers/subsystem/ticker.dm (hunk 6 above)
Comment (Xhuis) [inline code/controllers/subsystem/ticker.dm]: THE. KILL. STATS. ARE. DISABLED.

**Comment by Xhuis** [comment]
do it yourself

**Comment by Xhuis** [comment]
OK I'm not petulant anymore, so I'll explain. I'm tired of coding and can't muster the motivation to continue this.

**Comment by FlavoredCactus** [comment]
Didn't mean to strike a nerve, sorry mate.

**Comment by PKPenguin321** [comment]
flavoredcactus more like flavored dick

**Pull request closed without merging**